### PR TITLE
Fix JWT out-of-sync detection middleware and move it into LexQueries

### LINKED ISSUE #1180: JWT refresh middleware is buggy and out of date
**Describe the bug**
The middleware for detecting when to refresh JWT tokens is buggy and out of date.
- For projects: it refreshes too often, because the assumption that a user will only see projects they're a member of no longer holds due to non-confidential projects
- For orgs: it will essentially refresh precisely every time a user goes to the org list (which almost sounds like a good idea, but is certainly not the intention)

As a result:
- Visiting an org with non-confidential projects that the current user is not a member of will always refresh the token
- Unless the above is true ☝, the token will never be refreshed on an org page even if the user is added/removed from the org or their role is changed

**Expected behavior**
Our middleware should detect (whenever possible) actual permission changes and refresh the token accordingly.

### PR DESCRIPTION
**Pull request by myieye**
Resolves #1180 

This PR moves the JWT refreshing code into LexQueries.
It's no longer quite as generic as it was before (i.e. it won't automatically trigger based on GQL types), but it's much less magical, not so tricky to wire-up and easier to test.

The middleware was pretty buggy before. The unit tests should demonstrate that it handles every situation we're expecting our users to maybe land in.

Also, previously we used the old token for the scope of the request and only refreshed it at the end.
Now we look for updates up front and use the current state for the scope of the GQL Query.

Projecting a few Id's and UserId's lets us write less complicated code, so I thought it was worth it.

### CODE REVIEW

**Comment by myieye** [comment]
* **#1218** <a href="https://app.graphite.dev/github/pr/sillsdev/languageforge-lexbox/1218?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @myieye and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: frontend/src/routes/(authenticated)/project/[project_code]/+page.ts [hunk 1]
```diff
@@ -63,28 +60,26 @@ export async function load(event: PageLoadEvent) {
 						organizations {
 							id
 						}
-            ... on Project @include(if: $userIsMember) {
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply frontend/src/routes/(authenticated)/project/[project_code]/+page.ts:66]: alright I've fixed the issue so that members are now filtered (on the backend) if the user is not a member. However I'd like to keep this in place as there's no point in querying the users just for the backend to return an empty list when the user is not a member, although is that causing jwt refresh issues?

**Review by hahn-kev — CHANGES_REQUESTED** [review]
I'd like to fast track this as there's an open issue in production that it fixes.

File: frontend/src/routes/(authenticated)/project/[project_code]/+page.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (myieye) [reply frontend/src/routes/(authenticated)/project/[project_code]/+page.ts:66]: On the other hand, not requesting the users is sort of meaningless complexity.

Regarding JWT refreshing:
Yes, always selecting the users covers the small edge case of navigating directly to a project that you were added to, but is not yet in your token. That's what my [integration test](https://github.com/sillsdev/languageforge-lexbox/pull/1218/files#diff-6b87f4373687095a81128ab7fdbc2205b1172165a13b96b254134da86e1d8312R91) demonstrates.

It's an edge case, but I felt that handling it was essentially free, so I did. We're not excluding users from the GQL query for performance or security reasons. So, I think it makes sense to leave it.

File: backend/Testing/ApiTests/GqlMiddlewareTests.cs [hunk 2]
```diff
@@ -71,4 +86,49 @@ await Task.WhenAll(
...
+                projectByCode(code: "{{config.Code}}") {
+                    id
+                    name
+                    users {
+                        id
+                    }
+                }
+            }
+            """, expectGqlError: false); // we queried for users, so...
+        editorJwt.ShouldNotBe(_adminApiTester.CurrJwt); // token was updated
```
[the remark this answers is not in the document]
Comment (myieye) [reply backend/Testing/ApiTests/GqlMiddlewareTests.cs:132]: Totally, oops. ❤️

File: backend/LexBoxApi/Services/PermissionService.cs [hunk 3]
```diff
@@ -67,10 +67,11 @@ public async ValueTask AssertCanSyncProject(Guid projectId)
...
-    public async ValueTask<bool> CanViewProject(Guid projectId)
+    public async ValueTask<bool> CanViewProject(Guid projectId, LexAuthUser? overrideUser = null)
     {
-        if (User is not null && User.Role == UserRole.admin) return true;
-        if (User is not null && User.Projects.Any(p => p.ProjectId == projectId)) return true;
+        var user = overrideUser ?? User;
+        if (user is not null && user.Role == UserRole.admin) return true;
+        if (user is not null && user.Projects.Any(p => p.ProjectId == projectId)) return true;
         // Org admins can view all projects, even confidential ones
         if (await ManagesOrgThatOwnsProject(projectId)) return true;
```
[the remark this answers is not in the document]
Comment (myieye) [reply backend/LexBoxApi/Services/PermissionService.cs:76]: Good catch 👍

File: backend/LexBoxApi/Services/PermissionService.cs [hunk 4]
```diff
@@ -100,6 +100,8 @@ public async ValueTask<bool> CanViewProjectMembers(Guid projectId)
         if (User is not null && User.Role == UserRole.admin) return true;
         // Project managers can view members of their own projects, even confidential ones
         if (await CanManageProject(projectId)) return true;
+        // non members can't view project members
+        if (User?.IsProjectMember(projectId) != true) return false;
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/LexBoxApi/Services/PermissionService.cs:106]: I think I'd rather not, just because that's a much larger pool of users. Also this would be showing users who have been managers for a long time (I don't want it to be a consideration for making their project non confidential), since orgs are new and there will be less of them I feel better about making them public.

File: backend/LexCore/Auth/LexAuthUser.cs [hunk 5]
```diff
@@ -234,6 +234,15 @@ public ClaimsPrincipal GetPrincipal(string authenticationType)
             LexAuthConstants.EmailClaimType,
             LexAuthConstants.RoleClaimType));
     }
+
+    public bool IsProjectMember(Guid projectId, ProjectRole? role = null)
```
[the remark this answers is not in the document]
Comment (hahn-kev) [reply backend/LexCore/Auth/LexAuthUser.cs:238]: I'm fine with that, would you then remove the method from this class?

**Review by hahn-kev — APPROVED** [review]

File: backend/LexCore/Auth/LexAuthUser.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (myieye) [reply backend/LexCore/Auth/LexAuthUser.cs:238]: Yeah, I would.

As discussed, I'm going to merge this big PR and make the change in a follow up PR
